### PR TITLE
Completion fixes

### PR DESCRIPTION
**Pull request by paulvanbrenk**
This fixes 3 issues in the completion list, where we either didn't allow for a new identifier or shouldn't show the completion list at all.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
:-1:

File: src/services/services.ts [hunk 1]
```diff
@@ -2981,12 +2981,16 @@ module ts {
...
 
-              
                         case SyntaxKind.OpenParenToken:
                             return containingNodeKind === SyntaxKind.CallExpression               // func( |
                                 || containingNodeKind === SyntaxKind.Constructor                  // constructor( |
                                 || containingNodeKind === SyntaxKind.NewExpression                // new C(a|
-                                || containingNodeKind === SyntaxKind.ParenthesizedExpression;     // let x = (a|
+                                || containingNodeKind === SyntaxKind.ParenthesizedExpression      // let x = (a|
+                                || containingNodeKind === SyntaxKind.ParenthesizedType            // function F(pred: (a|
+                                || containingNodeKind == SyntaxKind.FunctionType;                 // function F(pred: (a| // todo: figure out why fourslash and the IDE don't agree on node kind here
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: this needs to be answered before checking in.

Comment (DanielRosenwasser) [inline src/services/services.ts]: We don't check in to git, we *merge*.

File: src/services/services.ts [hunk 2]
```diff
@@ -2981,12 +2981,15 @@ module ts {
...
                         case SyntaxKind.OpenParenToken:
                             return containingNodeKind === SyntaxKind.CallExpression               // func( |
                                 || containingNodeKind === SyntaxKind.Constructor                  // constructor( |
                                 || containingNodeKind === SyntaxKind.NewExpression                // new C(a|
-                                || containingNodeKind === SyntaxKind.ParenthesizedExpression;     // let x = (a|
+                                || containingNodeKind === SyntaxKind.ParenthesizedExpression      // let x = (a|
+                                || containingNodeKind === SyntaxKind.ParenthesizedType;            // function F(pred: (a
+
+                        case SyntaxKind.ColonToken:
+                            return containingNodeKind === SyntaxKind.Parameter;                   // function F(pred: a|
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: What about

* `var x: $`
* `function foo(): $`
* `interface I { method(): $ }`

etc?

File: src/services/services.ts [hunk 3]
```diff
@@ -2981,12 +2981,15 @@ module ts {
...
                                 || containingNodeKind === SyntaxKind.BinaryExpression;                      // let x = (a, |
 
-              
                         case SyntaxKind.OpenParenToken:
                             return containingNodeKind === SyntaxKind.CallExpression               // func( |
                                 || containingNodeKind === SyntaxKind.Constructor                  // constructor( |
                                 || containingNodeKind === SyntaxKind.NewExpression                // new C(a|
-                                || containingNodeKind === SyntaxKind.ParenthesizedExpression;     // let x = (a|
+                                || containingNodeKind === SyntaxKind.ParenthesizedExpression      // let x = (a|
+                                || containingNodeKind === SyntaxKind.ParenthesizedType;            // function F(pred: (a
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Align comment

File: src/services/services.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: This doesn't look right, what new identifier could you be writing here?

File: tests/cases/fourslash/completionListNewIdentifierFunctionDeclaration.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/// <reference path='fourslash.ts' />
+
+////function F(pref: (a/*1*/
+////
+////function F(pref: a/*2*/
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/completionListNewIdentifierFunctionDeclaration.ts]: Yeah, why is a new identifier allowed here?

Comment (DanielRosenwasser) [inline tests/cases/fourslash/completionListNewIdentifierFunctionDeclaration.ts]: Also this is better off split into two tests

Comment (paulvanbrenk) [inline tests/cases/fourslash/completionListNewIdentifierFunctionDeclaration.ts]: function F(pref: (a) => ... ){}

File: src/services/services.ts (hunk 2 above)
Comment (paulvanbrenk) [inline src/services/services.ts]: var x = (a) => ...

**Comment by DanielRosenwasser** [comment]
Add a test for when adding a parameter in the middle of an existing arrow type.

For instance
```TypeScript
var x: () => void
```

**Comment by DanielRosenwasser** [comment]
Also, what about tests with a second parameter in an arrow type?

```TypeScript
var frabblebabble: (rabble: number, list 
```

File: src/services/services.ts [hunk 5]
```diff
@@ -3138,6 +3141,9 @@ module ts {
                                 (previousToken.parent && previousToken.parent.parent &&
                                     previousToken.parent.parent.kind === SyntaxKind.ArrayBindingPattern);  // var [ ...z|
 
+                        case SyntaxKind.TypeKeyword:  // type htm|
```
Comment (yuit) [inline src/services/services.ts]: Move this together with SyntaxKind.ClassKeyword:, SyntaxKind.EnumKeyword etc.

**Comment by DanielRosenwasser** [comment]
:+1:

**Pull request merged**